<a href="https://colab.research.google.com/github/SimbakNetmind/series-de-tiempo/blob/02-M%C3%A9tricas-de-evaluaci%C3%B3n/01_M%C3%A9trica_de_evaluaci%C3%B3n(MAE).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Error Absoluto Medio (MAE)

El **MAE** (Mean Absolute Error) es una métrica fundamental para evaluar el desempeño de modelos de regresión.

### 📋 Definición y Características
* **Promedio de magnitudes:** Mide el promedio de las magnitudes absolutas de los errores.
* **Interpretación Directa:** Esta métrica es especialmente valorada por su interpretación directa, ya que se expresa en las **mismas unidades de la variable original**.
* **Robustez:** No penaliza de manera desproporcionada los errores grandes, lo que la hace una medida **robusta frente a valores atípicos** (outliers) moderados.
* **Importancia:** Es apropiada cuando todos los errores tienen la misma importancia.

> **Nota:** No distingue entre errores pequeños y grandes más allá de su magnitud absoluta.

---

### 🔢 Formulación Matemática

La fórmula para calcular el MAE es:

$$\text{MAE} = \frac{1}{T} \sum_{t=1}^{T} |Y_t - \hat{Y}_t|$$

Donde:
* $T$: Número total de observaciones.
* $Y_t$: Valor real.
* $\hat{Y}_t$: Valor predicho por el modelo.

# Requerimientos para la Aplicación de MAE

Para que el cálculo del **Error Absoluto Medio (MAE)** sea válido y proporcione información útil sobre el desempeño de un modelo, se deben cumplir los siguientes requisitos:

### 1. Naturaleza del Problema
* **Tarea de Regresión:** El MAE está diseñado exclusivamente para problemas de **regresión** (predicción de valores numéricos continuos). No es aplicable a tareas de clasificación donde las etiquetas son categóricas.
* **Escala de los Datos:** Las variables deben estar en una escala cuantitativa (intervalo o razón).

### 2. Estructura de los Datos
* **Pares de Comparación:** Se requiere un conjunto de datos que contenga:
    * **Valores Reales ($Y_t$):** Las etiquetas verdaderas u observaciones históricas.
    * **Valores Predichos ($\hat{Y}_t$):** Las estimaciones generadas por el modelo.
* **Mismo Tamaño ($T$):** Ambos vectores (reales y predicciones) deben tener exactamente la misma longitud o número de muestras.

### 3. Consistencia de Unidades
* **Unidades Homogéneas:** Tanto $Y_t$ como $\hat{Y}_t$ deben estar expresados en las **mismas unidades** de medida (por ejemplo, metros, dólares o grados Celsius).
* **Sin Transformaciones Inconsistentes:** Si los datos reales están en escala logarítmica, las predicciones también deben estarlo para que la resta $|Y_t - \hat{Y}_t|$ tenga sentido físico.

### 4. Supuestos sobre los Errores
* **Importancia Lineal:** Se debe asumir que el costo de un error aumenta de forma **lineal**. Es decir, un error de 10 unidades es exactamente el doble de grave que un error de 5 unidades.
* **Presencia de Outliers:** El MAE es ideal cuando el conjunto de datos tiene valores atípicos moderados y no queremos que estos "distorsionen" la métrica final (a diferencia de lo que ocurre con el MSE).

---

### Resumen Matemático de Componentes
Para ejecutar la fórmula $$\text{MAE} = \frac{1}{T} \sum_{t=1}^{T} |Y_t - \hat{Y}_t|$$ necesitas asegurar que:
1. $Y_t, \hat{Y}_t \in \mathbb{R}$
2. $T > 0$ (El conjunto no puede estar vacío).
3. No existan valores nulos (`NaN`) en ninguno de los dos vectores, ya que invalidarían la sumatoria.

### **Ejemplo de (MAE)**

### 🇩 Tabla de Métricas de Evaluación

| Día | Valor real | Pronóstico A | Pronóstico B | Pronóstico C |
|:---:|:----------:|:------------:|:------------:|:------------:|
| 1   | 95         | 92           | 94           | 90           |
| 2   | 97         | 93           | 96           | 101          |
| 3   | 96         | 98           | 95           | 92           |
| 4   | 98         | 94           | 99           | 103          |
| 5   | 100        | 96           | 100          | 95           |
| 6   | 102        | 99           | 101          | 107          |
| 7   | 101        | 97           | 100          | 96           |
| 8   | 103        | 105          | 103          | 99           |
| 9   | 105        | 100          | 104          | 110          |
| 10  | 104        | 101          | 105          | 98           |
| 11  | 106        | 110          | 105          | 112          |
| 12  | 108        | 104          | 108          | 102          |
| 13  | 107        | 109          | 106          | 113          |
| 14  | 109        | 104          | 110          | 103          |
| 15  | 110        | 106          | 109          | 116          |
| 16  | 112        | 115          | 112          | 106          |
| 17  | 111        | 108          | 112          | 117          |
| 18  | 113        | 110          | 114          | 107          |
| 19  | 114        | 118          | 113          | 120          |
| 20  | 115        | 111          | 115          | 109          |

---
## 📗 ***Instalación bibliotecas***

In [ ]:
!pip install numpy matplotlib pandas -q


## 🎯 ***Practica codigo (MAE)***

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Crear el DataFrame
data = {
    "Dia": np.arange(1, 21),
    "Valor_real": [95, 97, 96, 98, 100, 102, 101, 103, 105, 104,
                   106, 108, 107, 109, 110, 112, 111, 113, 114, 115],
    "Pronostico_A": [92, 93, 98, 94, 96, 99, 97, 105, 100, 101,
                     110, 104, 109, 104, 106, 115, 108, 110, 118, 111],
    "Pronostico_B": [94, 96, 95, 99, 100, 101, 100, 103, 104, 105,
                     105, 108, 106, 110, 109, 112, 112, 114, 113, 115],
    "Pronostico_C": [90, 101, 92, 103, 95, 107, 96, 99, 110, 98,
                     112, 102, 113, 103, 116, 106, 117, 107, 120, 109]
}



df = pd.DataFrame(data)


def mae(y, y_hat):
    return np.mean(np.abs(y - y_hat))

metricas = pd.DataFrame({
    "MAE": [
        mae(df["Valor_real"], df["Pronostico_A"]),
        mae(df["Valor_real"], df["Pronostico_B"]),
        mae(df["Valor_real"], df["Pronostico_C"])
    ]})

metricas


,MAE
0,3.50
1,0.75
2,5.40


### 🎯 Conclusión del Ejemplo MAE

Analizando los resultados de la métrica MAE para los diferentes pronósticos, se observa que:

*   **Pronóstico B** tiene el **MAE más bajo (0.75)**. Esto significa que, en promedio, las predicciones de este modelo están más cerca de los valores reales, con una desviación de 0.75 unidades.
*   **Pronóstico A** presenta un **MAE de 3.50**.
*   **Pronóstico C** tiene el **MAE más alto (5.40)**, lo que indica que sus predicciones son las que más se desvían de los valores reales.

**En resumen, el Pronóstico B es el modelo con mejor rendimiento** en este conjunto de datos, ya que un MAE menor indica una mayor precisión en las predicciones.

---
### 👽 Nota Extra (Modo Manual Ejemplo)

### 📊 **1. Cálculo con tus datos**

Primero tomamos el valor absoluto de cada resta:

|95 - 92|  = 3

|97 - 93|  = 4

|96 - 98|  = 2

|98 - 94|  = 4

|100 - 96| = 4

|102 - 99| = 3

|101 - 97| = 4

|103 - 105|= 2

|105 - 100|= 5

|104 - 101|= 3

|106 - 110|= 4

|108 - 104|= 4

|107 - 109|= 2

|109 - 104|= 5

|110 - 106|= 4

|112 - 115|= 3

|111 - 108|= 3

|113 - 110|= 3

|114 - 118|= 4

|115 - 111|=

### **3. Sumamos los errores absolutos**

3 + 4 + 2 + 4 + 4 + 3 + 4 + 2 + 5 + 3 + 4 + 4 + 2 + 5 + 4 + 3 + 3 + 3 + 4 + 4 = 70

### 3. **Calculamos el MA**

Hay 20 datos, entonces:

###✅  **Resultado final**

👉 ***MAE = 3.5***